core/src/sql/db_connection_pool/adbcpool.rs

```diff
@@ -0,0 +1,249 @@
+        self.max_size = size;
+        self
+    }
+
+    pub fn with_min_idle(mut self, size: Option<u32>) -> Self {
+        self.min_idle = size;
+        self
+    }
+
+    pub fn build(self) -> Result<ADBCPool<D>> {
+        let mut pool_builder = r2d2::Pool::builder();
+
+        if let Some(size) = self.max_size {
+            pool_builder = pool_builder.max_size(size);
+        }
+        if self.min_idle.is_some() {
+            pool_builder = pool_builder.min_idle(self.min_idle);
+        }
+
+        // let opts = ;
+        let database = self.database;
+        let manager: AdbcConnectionManager<D>;
+        if let Some(opts) = self.connection_options {
+            manager = AdbcConnectionManager::with_options(database, opts);
+        } else {
+            manager = AdbcConnectionManager::new(database);
+        }
+        let pool = Arc::new(pool_builder.build(manager).context(ConnectionPoolSnafu)?);
+        Ok(ADBCPool { pool })
+    }
+}
+
+#[derive(Clone)]
+pub struct ADBCPool<D>
+where
+    D: Database + Send + 'static,
+    D::ConnectionType: Connection + Send + Sync,
+{
+    pool: Arc<r2d2::Pool<AdbcConnectionManager<D>>>,
+}
+
+impl<D> std::fmt::Debug for ADBCPool<D>
+where
+    D: Database + Send + 'static,
+    D::ConnectionType: Connection + Send + Sync,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ADBCPool")
+    }
+}
+
+impl<D> ADBCPool<D>
+where
+    D: Database + Send + 'static,
+    D::ConnectionType: Connection + Send + Sync,
+{
+    pub fn new(db: D, conn_options: Option<HashMap<String, String>>) -> Result<Self> {
+        let builder = AdbcConnectionPoolBuilder::new(db);
+        let builder = if let Some(opts) = conn_options {
+            builder.with_conn_options(opts)
+        } else {
+            builder
+        };
+        builder.build()
+    }
+
+    pub fn connect_sync(
+        self: Arc<Self>,
+    ) -> Result<Box<dyn DbConnection<r2d2::PooledConnection<AdbcConnectionManager<D>>, RecordBatch>>>
+    {
+        let pool = Arc::clone(&self.pool);
+        let conn: r2d2::PooledConnection<AdbcConnectionManager<D>> =
+            pool.get().context(ConnectionPoolSnafu)?;
+
+        Ok(Box::new(AdbcDbConnection::new(conn)))
+    }
+}
+
+#[async_trait]
+impl<D> DbConnectionPool<r2d2::PooledConnection<AdbcConnectionManager<D>>, RecordBatch>
+    for ADBCPool<D>
+where
+    D: Database + Send + 'static,
+    D::ConnectionType: Connection + Send + Sync,
+{
+    async fn connect(
+        &self,
+    ) -> Result<Box<dyn DbConnection<r2d2::PooledConnection<AdbcConnectionManager<D>>, RecordBatch>>>
+    {
+        let pool = Arc::clone(&self.pool);
+        let conn: r2d2::PooledConnection<AdbcConnectionManager<D>> =
+            pool.get().context(ConnectionPoolSnafu)?;
+
+        Ok(Box::new(AdbcDbConnection::new(conn)))
+    }
+
+    fn join_push_down(&self) -> JoinPushDown {
+        // TODO: make this configurable
```

Comment:
This could probably be set to a hash of the connection information, but it can come in a follow up.